Mathlib/Tactic/Choose.lean

```diff
@@ -0,0 +1,265 @@
+if h : Nonempty α then f (choice h) else choice ‹_›
+
+theorem sometimes_eq {p : Prop} {α} [Nonempty α] (f : p → α) (a : p) : sometimes f = f a :=
+dif_pos ⟨a⟩
+
+theorem sometimes_spec {p : Prop} {α} [Nonempty α]
+  (P : α → Prop) (f : p → α) (a : p) (h : P (f a)) : P (sometimes f) :=
+(sometimes_eq f a).symm ▸ h
+
+end function
+
+namespace Tactic
+
+namespace Choose
+
+/-- Given `α : Sort u`, `nonemp : Nonempty α`, `p : α → Prop`, a context of free variables
+`ctxt`, and a pair of an element `val : α` and `spec : p val`,
+`mk_sometimes u α nonemp p ctx (val, spec)` produces another pair `val', spec'`
+such that `val'` does not have any free variables from elements of `ctxt` whose types are
+propositions. This is done by applying `function.sometimes` to abstract over all the propositional
+arguments. -/
+def mk_sometimes (u : Level) (α nonemp p : Expr) :
+  List Expr → Expr × Expr → MetaM (Expr × Expr)
+| [], (val, spec) => pure (val, spec)
+| (e :: ctxt), (val, spec) => do
+  let (val, spec) ← mk_sometimes u α nonemp p ctxt (val, spec)
+  let t ← inferType e
+  let b ← isProp t
+  if b then do
+    let val' ← mkLambdaFVars #[e] val
+    pure
+      (mkApp4 (Expr.const ``function.sometimes [Level.zero, u]) t α nonemp val',
+      mkApp7 (Expr.const ``function.sometimes_spec [u]) t α nonemp p val' e spec)
+  else pure (val, spec)
+
+/-- Results of searching for nonempty instances,
+to eliminate dependencies on propositions (`choose!`).
+`success` means we found at least one instance;
+`failure ts` means we didn't find instances for any `t ∈ ts`.
+(`failure []` means we didn't look for instances at all.)
+
+Rationale:
+`choose!` means we are expected to succeed at least once
+in eliminating dependencies on propositions.
+-/
+inductive ElimStatus
+| success
+| failure (ts : List Expr)
+
+/-- Combine two statuses, keeping a success from either side
+or merging the failures. -/
+def ElimStatus.merge : ElimStatus → ElimStatus → ElimStatus
+| success, _ => success
+| _, success => success
+| failure ts₁, failure ts₂ => failure (ts₁ ++ ts₂)
+
+def getBinderName (e : Expr) : MetaM (Option Name) := do
+  match ← withReducible (whnf e) with
+  | .forallE (binderName := n) .. | .lam (binderName := n) .. => pure (some n)
+  | _ => pure none
+
+def mkFreshNameFrom (orig base : Name) : CoreM Name :=
+  if orig = `_ then mkFreshUserName base else pure orig
+
+def SourceInfo.fromRef' (ref : Syntax) (synthetic := true) : SourceInfo :=
+  match ref.getPos?, ref.getTailPos? with
+  | some pos, some tailPos =>
+    if synthetic then .synthetic pos tailPos
+    else .original "".toSubstring pos "".toSubstring tailPos
+  | _,        _            => .none
+
+def addLocalVarInfoForBinderIdent (fvar : Expr) : TSyntax ``binderIdent → TermElabM Unit
+| `(binderIdent| $n:ident) =>
+  Elab.Term.addLocalVarInfo n fvar
+| tk => do
+  let stx := mkNode ``Parser.Term.hole #[Syntax.atom (SourceInfo.fromRef' tk false) "_"] -- HACK
+  Elab.Term.addLocalVarInfo stx fvar
```

Comment:
Where should these be moved?

Reply:
it's not super important, it will get moved around anyway later. Possible locations include: `Mathlib.Lean` for stuff that "should have been in core already", `Mathlib.Util` for general utility definitions, and `Mathlib.Tactic.Core` for stuff used in most tactics